---
title: "kegg_modules_rawdOD_070221"
author: "Elena"
date: "7/2/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(pheatmap)
library(viridis)
library(readxl)
library(here)
library(ggpubr)
library(clusterProfiler)
library(pathview)
```

## Growth Profile Data ##

```{r}
# see format_growth_profile.Rmd for explanation of code
source(here('growthProfiles.R'))
rm(cleanGE, dOD)

# Luca's gene annotation and KEGG module distances (copied original tsv into an excel file)
ann <- readxl::read_xlsx(here('data', 'ann_master.xlsx'), col_names = TRUE)

# subset gene annotations depending on whether they have EC numbers
# "metabolic" genes
met <- ann %>% 
  filter(!KO == "NA") %>% 
  select(strain.nickname, KO, gene) %>%
  rename(strain = strain.nickname)

# wide matrix metabolic genes
# binary matrix gene names for annotations WITH EC numbers
geneWide <- met %>% select(strain, KO) %>% distinct() %>% 
  mutate(genfill = 1) %>%
  pivot_wider(names_from = KO, values_from = genfill) %>%
  replace(is.na(.), 0)
rm(met)

# always end up needing tax info
tax <- readxl::read_xlsx(here('data', 'strain_tax copy.xlsx')) %>%
  dplyr::select(strain, phylum, class, order, family, genus, gnm_GFC)

# single gene correlation values from pointBiserial.Rmd
rpb_all <- readxl::read_xlsx('all_high_r_ko_011722.xlsx')
```

# Pathway Mapping

```{r eval=FALSE}
#packageurl <- "https://cran.r-project.org/src/contrib/Archive/rvcheck/rvcheck_0.1.8.tar.gz"
#install.packages(packageurl, repos=NULL, type="source")
#remove.packages("clusterProfiler")
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("clusterProfiler")
#BiocManager::install("DO.db", type = "source")
#BiocManager::install("GO.db", type = "source")
#BiocManager::install("pathview")
#BiocManager::install("org.Hs.eg.db", type = "source")
```

```{r}
# map high r KOs to KEGG using clusterprofiler
rko <- rpb_all %>%
  rename("KO" = gene) %>%
  filter(p.adj < .05) %>% # filter to look at stat sig KOs only
  select(medium, KO, pbCorr, p.adj) 

medlist <- unique(rko$medium)

for (i in medlist) {
  dat <- rko %>%
    filter(medium == i)
  # map high-r genes to KEGG
  res <- enrichKEGG(gene = dat$KO,
                      organism = 'ko',
                      minGSSize = 10,
                      maxGSSize = 500,
                      pvalueCutoff = 0.05,
                      qvalueCutoff = 0.2,
                      pAdjustMethod = "BH")
  # filter pathways to which fewer than 3 genes mapped
  resdf <- gsfilter(res, by = "Count", min = 3)@result %>% 
    mutate(medium = i)
 
  assign(paste0('enrpath_', 'clst', i), resdf)
  
}

rm(i, dat, res, resdf)
```

```{r}
## print KEGG maps of all the enriched pathways ##
# you need to do the kegg enrichment with KEGGprofile above
# pathview prints the reference (uncolored) pathway as well

# function to print KEGG maps
# x is the string for name of medium to search
# y is the string for the pathway
printEnrichKegg <- function(x, y) {
  geno <- rko %>% filter(medium == x) %>% dplyr::select(KO, pbCorr)
  g <- geno$pbCorr
  names(g) <- geno$KO
  pv.out <- pathview(gene.data = g, 
                   pathway.id = y, # pathway id taken from clusterprofiler results
                   species = "ko", # reference
                   out.suffix = paste0(Sys.Date(), '_', x), 
                   kegg.native = T,
                   same.layer = F)
}

## print KEGG pathways identified as significantly enriched by clusterprofiler with high-r KOs ##
for (m in medlist) {
  dat <- eval(as.name(paste0('enrpath_', 'clst', m)))
  for (i in 1:nrow(dat)) {
    printEnrichKegg(m, dat[i,1])
  }
}

rm(i, m)
```

## Heatmap

```{r}
# significantly enriched kegg pathways from clusterprofiler
enrpathCP <- tibble()

for (m in medlist) {
  dat <- eval(as.name(paste0('enrpath_', 'clst', m))) %>%
    mutate(medium = m)
  enrpathCP <- bind_rows(enrpathCP, dat)
}

# the pathways below are not informative, too general
rmPath <- c("Carbon metabolism", "Fatty acid biosynthesis", "Degradation of Aromatic Compounds", 
            "Biosynthesis of amino acids", "Biosynthesis of cofactors", "Biosynthesis of nucleotide sugars", 
            "Cell cycle - Caulobacter", "Fatty acid metabolism", "Degradation of aromatic compounds")

enrpathCP <- enrpathCP %>%
  select(medium, ID, Description, GeneRatio, BgRatio, Count, p.adjust) %>%
  filter(!Description %in% rmPath & p.adjust < .05 ) %>%
  mutate(medium = factor(medium, levels = c('difcoMB', 'HMBcmpt', 'HMBpep', 'HMBaa', 'HMBlips', 'HMBoligo', 
                                    'HMBorg', 'HMBntrl', 'HMBamisug', 'HMBacdsug', 'HMB--')))

# write to file
#writexl::write_xlsx(enrpathCP, 'enrichedKEGGpaths_011721.xlsx')
```

```{r}
# enriched pathways per medium
enrpathCP %>%
  group_by(medium) %>%
  summarise(n())

enrpathCP %>%
  ggplot(aes(x = medium)) +
  geom_bar() +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))

enrpathCP %>%
  ggplot(aes(x = medium, y = p.adjust)) +
  geom_point() +
  geom_boxplot(alpha = .5, outlier.size = -1) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
  
```

I annotated all of the pathway names in sig_enriched_pathway_names.xlsx by hand and checked all of the KOs mapping to those pathways. I then added a tab to the file with the classification according to BRITE hierarchies, with a couple noted exceptions. I excluded HMBcmpt and difcoMB, pathways not significantly enriched for the other media, and pathways that were too general to be informative (eg "carbon metabolism").

```{r}
# classification for all hand-checked enriched pathways
pathClass <- readxl::read_xlsx('sig_enriched_pathway_names.xlsx', sheet = "classification")

cpathClass <- enrpathCP %>%
  left_join(pathClass, by = "ID") %>%
  select(ID, Description.x, cat1, cat2, medium) %>%
  arrange(cat1, cat2, Description.x) %>%
  rename(Description = Description.x)
```

```{r}
cpathClass_wide <- cpathClass %>%
  select(Description, medium) %>%
  mutate(present = 1) %>%
  pivot_wider(names_from = Description, values_from = present) %>%
  arrange(factor(medium, levels = c('difcoMB', 'HMBcmpt', 'HMBpep', 'HMBaa', 'HMBlips', 'HMBoligo', 
                                      'HMBorg', 'HMBntrl', 'HMBamisug', 'HMBacdsug', 'HMB--')))

cpathClass_wide[is.na(cpathClass_wide)] <- 0

z <- select(cpathClass_wide, -medium)
rownames(z) <- cpathClass_wide$medium

# annotations
c1_rownames <- cpathClass %>%
  select(Description, cat1, cat2) %>%
  distinct() 
c1 <- data.frame(category1 = factor(c1_rownames$cat1), # first col must be a factor, apparently?
                 category2 = c1_rownames$cat2)
rownames(c1) <- colnames(z)
# alternative annotation order
c2 <- data.frame(category1 = factor(c1_rownames$cat2), # first col must be a factor, apparently?
                 category2 = c1_rownames$cat1)
rownames(c2) <- colnames(z)

# annotation colors
cols1 <- colorRampPalette(turbo(17))
cols2 <- colorRampPalette(plasma(4))
mycolors <- list(category1 = cols1(length(unique(c2$category1))),
                 category2 = cols2(length(unique(c2$category2)))
)
names(mycolors$category1) <- unique(c2$category1)
names(mycolors$category2) <- unique(c2$category2)

# heatmap
pheatmap::pheatmap(t(z),
                   cluster_rows = FALSE,
                   cluster_cols = FALSE,
                   color = mako(100),
                   annotation_row = c2,
                   annotation_colors = mycolors,
                   annotation_names_row = FALSE,
                   border_color = "grey60",
                   gaps_row = c(5, 9, 13, 22, 23, 37, 42, 43, 51, 55, 57, 59),
                   fontsize = 11)
          

```


















































